src/utilities/tree_interface.jl

```diff
@@ -38,3 +38,97 @@ function AbstractTrees.printnode(io::IO, node::Constant)
 end
 
 AbstractTrees.printnode(io::IO, node::AbstractVariable) = summary(io, node)
+
+mutable struct Counts
+    seen::Base.IdSet
+    n_variables::Int
+    n_scalar_variables::Int
+    n_constraints::Int
+    n_scalar_constraints::Int
+    n_atoms::Int
+    n_nonzeros::Int
+    Counts() = new(Base.IdSet(), 0, 0, 0, 0, 0, 0)
+end
+
+function Counts(p::Problem)
+    counts = Counts()
+    _add_to_problem_count(counts, p)
+    counts.n_atoms -= 1  # Don't count p as an atom
+    return counts
+end
+
+function _add_to_problem_count(counts::Counts, node::AbstractExpr)
+    if node in counts.seen
+        return
+    end
+    for n in AbstractTrees.PreOrderDFS(node)
+        if n !== node && !(n in counts.seen)
+            _add_to_problem_count(counts, n)
+        end
+    end
+    counts.n_atoms += 1
+    push!(counts.seen, node)
+    return
+end
+
+function _add_to_problem_count(counts::Counts, node::AbstractVariable)
+    if node in counts.seen
+        return
+    end
+    counts.n_variables += 1
+    counts.n_scalar_variables += (iscomplex(node) ? 2 : 1) * length(node)
+    push!(counts.seen, node)
+    for c in get_constraints(node)
+        _add_to_problem_count(counts, c)
```

Comment:
I wonder if one could have a constraint referenced in multiple places?
```suggestion
       if !(c in counts.seen)
            _add_to_problem_count(counts, c)
        end
```